jib-maven-plugin/src/test/java/com/google/cloud/tools/jib/maven/skaffold/SyncMapMojoTest.java

```diff
@@ -0,0 +1,160 @@
+  }
+
+  private static String getSyncMapJson(Path projectRoot, String module)
+      throws VerificationException, IOException {
+    Path logFile = runBuild(projectRoot, module, null);
+    List<String> outputLines = Files.readAllLines(logFile, StandardCharsets.UTF_8);
+    Assert.assertEquals(3, outputLines.size()); // we expect ["\n", "BEGIN JIB JSON", "<sync-json>"]
+    return outputLines.get(2); // this is the JSON output
+  }
+
+  private static void assertFilePaths(Path src, AbsoluteUnixPath dest, FileTemplate template) {
+    Assert.assertEquals(src.toString(), template.getSrc());
+    Assert.assertEquals(dest.toString(), template.getDest());
+  }
+
+  @Test
+  public void testSyncMapMojo_simpleTestProjectOutput() throws IOException, VerificationException {
+    Path projectRoot = simpleTestProject.getProjectRoot();
+    String json = getSyncMapJson(projectRoot, null);
+    SkaffoldSyncMapTemplate parsed = SkaffoldSyncMapTemplate.from(json);
+
+    List<FileTemplate> generated = parsed.getGenerated();
+    Assert.assertEquals(2, generated.size());
+    assertFilePaths(
+        projectRoot.resolve("target/classes/world"),
+        AbsoluteUnixPath.get("/app/resources/world"),
+        generated.get(0));
+    assertFilePaths(
+        projectRoot.resolve("target/classes/com/test/HelloWorld.class"),
+        AbsoluteUnixPath.get("/app/classes/com/test/HelloWorld.class"),
+        generated.get(1));
+
+    List<FileTemplate> direct = parsed.getDirect();
+    Assert.assertEquals(2, direct.size());
+    assertFilePaths(
+        projectRoot.resolve("src/main/jib-custom/bar/cat"),
+        AbsoluteUnixPath.get("/bar/cat"),
+        direct.get(0));
+    assertFilePaths(
+        projectRoot.resolve("src/main/jib-custom/foo"),
+        AbsoluteUnixPath.get("/foo"),
+        direct.get(1));
+  }
+
+  @Test
+  public void testSyncMapMojo_multiProjectOutput() throws IOException, VerificationException {
+    Path projectRoot = multiTestProject.getProjectRoot();
+    Path m2 = Paths.get(System.getProperty("user.home")).resolve(".m2").resolve("repository");
+    String json = getSyncMapJson(projectRoot, "complex-service");
+    SkaffoldSyncMapTemplate parsed = SkaffoldSyncMapTemplate.from(json);
+
+    List<FileTemplate> generated = parsed.getGenerated();
+    Assert.assertEquals(2, generated.size());
+    assertFilePaths(
+        projectRoot.resolve("lib/target/lib-1.0.0.TEST-SNAPSHOT.jar"),
+        AbsoluteUnixPath.get("/app/libs/lib-1.0.0.TEST-SNAPSHOT.jar"),
+        generated.get(0));
+    assertFilePaths(
+        projectRoot.resolve("complex-service/target/classes/com/test/HelloWorld.class"),
+        AbsoluteUnixPath.get("/app/classes/com/test/HelloWorld.class"),
+        generated.get(1));
+
+    List<FileTemplate> direct = parsed.getDirect();
+    Assert.assertEquals(1, direct.size());
+    assertFilePaths(
+        m2.resolve(
+            "com/google/cloud/tools/tiny-test-lib/0.0.1-SNAPSHOT/tiny-test-lib-0.0.1-SNAPSHOT.jar"),
```

Comment:
Interesting. `runBuild()` does `mvn package` only, and I know this custom-built JAR is under `jib-maven-plugin/build/resources/test/maven/projects/multi/complex-service/local-m2-repo/com/google/cloud/tools/tiny-test-lib`. Who installs this into `<user.home>/.m2/repository`? I just can't understand how this ended up in the user's m2 repo.

And does this mean that we will sync a SNAPSHOT dependency (which is not a project dependency) too? When the SNAPSHOT dependency in a local Maven repo is updated?

Reply:
Honestly, I think I was on autopilot when I was testing this, I have no idea how this works. Lemme go look at it again.

Reply:
So it turns out that maven treats the the local-m2-repo as a remote repository because we define it as:
```
  <repositories>
    <repository>
      <id>local-m2-repo</id>
      <url>file:${project.basedir}/local-m2-repo</url>
    </repository>
  </repositories>
```

and so it's copied into the .m2 when the build happens from the remote "local-m2-repo", I'll clean this up in a followup.